# 01 EDA: разведочный анализ логов Ozon Similar Products

Цель: понять, готовы ли текущие логи к preprocessing, sessionization и первому co-visitation baseline.

## 1. Импорты и пути

Таблица пользовательских действий большая, поэтому ноутбук не загружает весь лог в память. Используем parquet metadata, lazy scans, выбор нужных колонок и дневные проходы для дорогих проверок.

In [1]:
import sys
from pathlib import Path

PROJECT_ROOT = Path.cwd()
if PROJECT_ROOT.name == "notebooks":
    PROJECT_ROOT = PROJECT_ROOT.parent
if str(PROJECT_ROOT / "src") not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT / "src"))

PROJECT_ROOT

WindowsPath('C:/Users/User/Projects/OZON-Similar-products')

In [ ]:
import polars as pl

from ozon_similar_products.data import load_products
from ozon_similar_products.data.loaders import (
    find_parquet_payload_dir,
    get_path_from_config,
    load_configs,
)
from ozon_similar_products.data.profiling import (
    action_profile,
    null_profile,
    parquet_dataset_overview,
    parquet_partition_profile,
    partition_row_counts,
    schema_overview,
)
from ozon_similar_products.data.schemas import KNOWN_ACTION_TYPES
from ozon_similar_products.features.item_popularity import (
    load_event_weights,
    weighted_item_popularity,
)
from ozon_similar_products.preprocessing.session_checks import time_diff_summary

pl.Config.set_tbl_rows(30)
pl.Config.set_tbl_cols(30)

PROJECT_CONFIG = load_configs(project_root=PROJECT_ROOT)
DATA_CONFIG = PROJECT_CONFIG["data"]

PRODUCTS_BASE_DIR = get_path_from_config(
    PROJECT_CONFIG,
    section="data",
    key="product_information_dir",
)
USER_ACTIONS_BASE_DIR = get_path_from_config(
    PROJECT_CONFIG,
    section="data",
    key="user_actions_dir",
)

PRODUCTS_ROOT = find_parquet_payload_dir(
    base_dir=PRODUCTS_BASE_DIR,
    payload_root_names=DATA_CONFIG["product_information"]["payload_root_names"],
    parquet_glob=DATA_CONFIG["product_information"]["parquet_glob"],
)
ACTIONS_ROOT = find_parquet_payload_dir(
    base_dir=USER_ACTIONS_BASE_DIR,
    payload_root_names=DATA_CONFIG["user_actions"]["payload_root_names"],
    parquet_glob=DATA_CONFIG["user_actions"]["parquet_glob"],
)

PRODUCTS_ROOT, ACTIONS_ROOT

## 2. Обзор справочника товаров

Справочник товаров достаточно маленький, чтобы загрузить его полностью. Проверяем `item_id`, полноту метаданных и пригодность признаков для будущего fallback и business rules.

In [3]:
products = load_products(config=PROJECT_CONFIG)

product_schema = schema_overview(products)
product_nulls = null_profile(products)
product_summary = pl.DataFrame(
    {
        "метрика": [
            "строк",
            "уникальных item_id",
            "строк с дублями item_id",
            "уникальных category_id",
            "уникальных category_name",
            "уникальных brand",
            "уникальных type",
        ],
        "значение": [
            products.height,
            products.select(pl.col("item_id").n_unique()).item(),
            products.height - products.select(pl.col("item_id").n_unique()).item(),
            products.select(pl.col("category_id").n_unique()).item(),
            products.select(pl.col("category_name").n_unique()).item(),
            products.select(pl.col("brand").n_unique()).item(),
            products.select(pl.col("type").n_unique()).item(),
        ],
    }
)

product_summary, product_schema, product_nulls

(shape: (7, 2)
 ┌──────────────────────────┬──────────┐
 │ метрика                  ┆ значение │
 │ ---                      ┆ ---      │
 │ str                      ┆ i64      │
 ╞══════════════════════════╪══════════╡
 │ строк                    ┆ 130035   │
 │ уникальных item_id       ┆ 130035   │
 │ строк с дублями item_id  ┆ 0        │
 │ уникальных category_id   ┆ 834      │
 │ уникальных category_name ┆ 796      │
 │ уникальных brand         ┆ 6682     │
 │ уникальных type          ┆ 2184     │
 └──────────────────────────┴──────────┘,
 shape: (6, 2)
 ┌───────────────┬────────┐
 │ column        ┆ dtype  │
 │ ---           ┆ ---    │
 │ str           ┆ str    │
 ╞═══════════════╪════════╡
 │ item_id       ┆ Int32  │
 │ name          ┆ String │
 │ brand         ┆ String │
 │ type          ┆ String │
 │ category_id   ┆ Int32  │
 │ category_name ┆ String │
 └───────────────┴────────┘,
 shape: (6, 4)
 ┌───────────────┬───────────┬────────────┬────────────┐
 │ column        ┆ row_coun

In [4]:
top_categories = (
    products.group_by(["category_id", "category_name"])
    .agg(pl.len().alias("items"))
    .sort("items", descending=True)
    .head(30)
)

top_brands = (
    products.group_by("brand")
    .agg(pl.len().alias("items"))
    .sort("items", descending=True)
    .head(30)
)

top_types = (
    products.group_by("type")
    .agg(pl.len().alias("items"))
    .sort("items", descending=True)
    .head(30)
)

top_categories, top_brands, top_types

(shape: (30, 3)
 ┌─────────────┬─────────────────────────────────┬───────┐
 │ category_id ┆ category_name                   ┆ items │
 │ ---         ┆ ---                             ┆ ---   │
 │ i32         ┆ str                             ┆ u32   │
 ╞═════════════╪═════════════════════════════════╪═══════╡
 │ 54          ┆ Наполнители                     ┆ 7698  │
 │ 445         ┆ Сухие корма                     ┆ 3287  │
 │ 347         ┆ Для цветного белья              ┆ 3142  │
 │ 754         ┆ Сухие корма                     ┆ 2954  │
 │ 771         ┆ Влажные корма                   ┆ 2515  │
 │ 671         ┆ Блочные                         ┆ 2211  │
 │ 661         ┆ Специальные чистящие средства   ┆ 1939  │
 │ 731         ┆ Подгузники-трусики              ┆ 1860  │
 │ 849         ┆ Зубные пасты, гели              ┆ 1805  │
 │ 301         ┆ Смартфоны                       ┆ 1739  │
 │ 817         ┆ Для белого белья                ┆ 1538  │
 │ 284         ┆ Электрические чайники  

## 3. Логи действий: parquet metadata и дневные партиции

Этот блок читает parquet metadata, а не сами строки. Он должен выполняться быстро даже на полном датасете.

In [5]:
actions_overview = parquet_dataset_overview(ACTIONS_ROOT)
actions_file_profile = parquet_partition_profile(ACTIONS_ROOT)
date_action_counts = partition_row_counts(ACTIONS_ROOT)

date_summary = (
    date_action_counts.group_by("date")
    .agg(
        pl.col("rows").sum().alias("rows"),
        pl.col("files").sum().alias("files"),
        pl.col("file_size_bytes").sum().alias("file_size_bytes"),
    )
    .sort("date")
)

action_rows_from_metadata = (
    date_action_counts.group_by("action_type")
    .agg(pl.col("rows").sum().alias("rows"))
    .with_columns((pl.col("rows") / pl.col("rows").sum()).alias("share"))
    .sort("rows", descending=True)
)

actions_overview, date_summary.head(), date_summary.tail(), action_rows_from_metadata

(shape: (1, 3)
 ┌───────┬───────────┬─────────────────┐
 │ files ┆ rows      ┆ file_size_bytes │
 │ ---   ┆ ---       ┆ ---             │
 │ u32   ┆ i64       ┆ i64             │
 ╞═══════╪═══════════╪═════════════════╡
 │ 305   ┆ 561836992 ┆ 11761209986     │
 └───────┴───────────┴─────────────────┘,
 shape: (5, 4)
 ┌────────────┬──────────┬───────┬─────────────────┐
 │ date       ┆ rows     ┆ files ┆ file_size_bytes │
 │ ---        ┆ ---      ┆ ---   ┆ ---             │
 │ str        ┆ i64      ┆ u32   ┆ i64             │
 ╞════════════╪══════════╪═══════╪═════════════════╡
 │ 2024-03-01 ┆ 10446059 ┆ 5     ┆ 223224248       │
 │ 2024-03-02 ┆ 11454573 ┆ 5     ┆ 244129508       │
 │ 2024-03-03 ┆ 11127161 ┆ 5     ┆ 238471083       │
 │ 2024-03-04 ┆ 9638934  ┆ 5     ┆ 208663397       │
 │ 2024-03-05 ┆ 10859997 ┆ 5     ┆ 233281986       │
 └────────────┴──────────┴───────┴─────────────────┘,
 shape: (5, 4)
 ┌────────────┬─────────┬───────┬─────────────────┐
 │ date       ┆ rows    ┆ files

In [6]:
partition_health = pl.DataFrame(
    {
        "метрика": [
            "минимальная дата",
            "максимальная дата",
            "дней",
            "parquet-файлов",
            "строк",
            "минимум строк за день",
            "максимум строк за день",
            "найденные известные action_type",
        ],
        "значение": [
            str(date_summary.select(pl.col("date").min()).item()),
            str(date_summary.select(pl.col("date").max()).item()),
            str(date_summary.height),
            str(actions_overview["files"].item()),
            str(actions_overview["rows"].item()),
            str(date_summary.select(pl.col("rows").min()).item()),
            str(date_summary.select(pl.col("rows").max()).item()),
            ", ".join(sorted(set(action_rows_from_metadata["action_type"].to_list()) & set(KNOWN_ACTION_TYPES))),
        ],
    }
)

partition_health

метрика,значение
str,str
"""минимальная дата""","""2024-03-01"""
"""максимальная дата""","""2024-04-30"""
"""дней""","""61"""
"""parquet-файлов""","""305"""
"""строк""","""561836992"""
"""минимум строк за день""","""7431439"""
"""максимум строк за день""","""12215617"""
"""найденные известные action_typ…","""click, favorite, search, to_ca…"


## 4. Lazy scan и схема логов

Дальше используем lazy scans и выбираем только те колонки, которые нужны для конкретной проверки.

In [7]:
user_actions_lf = pl.scan_parquet(str(ACTIONS_ROOT / "**" / "*.parquet"), hive_partitioning=True)
actions_schema = schema_overview(user_actions_lf)
actions_schema

column,dtype
str,str
"""user_id""","""Int32"""
"""date""","""Date"""
"""timestamp""","""Datetime(time_unit='ns', time_…"
"""action_type""","""String"""
"""widget_name""","""String"""
"""search_query""","""String"""
"""item_id""","""Int32"""


## 5. Анализ типов событий, пропусков и widget-контекста

Этот проход читает только ключевые колонки. У событий `search` поле `item_id` может быть пустым; это нормально, такие события нужно обрабатывать отдельно от прямых товарных взаимодействий.

`widget_name` проверяем отдельным разрезом, потому что одинаковый `action_type` может приходить из разных зон интерфейса. Для MVP direct item events остаются в графе, а widget distribution становится обязательной диагностикой перед baseline.

In [ ]:
key_action_columns = ["user_id", "timestamp", "action_type", "widget_name", "search_query", "item_id"]
actions_key_lf = user_actions_lf.select(key_action_columns)
direct_action_types = ["view", "click", "favorite", "to_cart"]

actions_by_type = action_profile(actions_key_lf).rename(
    {"users": "unique_users", "items": "unique_items"}
)
actions_nulls = null_profile(actions_key_lf)

widget_action_profile = (
    actions_key_lf
    .group_by(["action_type", "widget_name"])
    .agg(
        pl.len().alias("rows"),
        pl.col("user_id").n_unique().alias("unique_users"),
        pl.col("item_id").drop_nulls().n_unique().alias("unique_items"),
    )
    .sort(["action_type", "rows"], descending=[False, True])
    .collect()
)

direct_widget_profile = (
    actions_key_lf
    .filter(pl.col("action_type").is_in(direct_action_types))
    .filter(pl.col("item_id").is_not_null())
    .group_by("widget_name")
    .agg(
        pl.len().alias("rows"),
        pl.col("user_id").n_unique().alias("unique_users"),
        pl.col("item_id").n_unique().alias("unique_items"),
    )
    .with_columns((pl.col("rows") / pl.col("rows").sum()).alias("share"))
    .sort("rows", descending=True)
    .collect()
)

actions_by_type, actions_nulls, widget_action_profile, direct_widget_profile

In [ ]:
missing_value_decisions = pl.DataFrame(
    {
        "условие": [
            "action_type = search and item_id is null",
            "action_type in view/click/favorite/to_cart and item_id is null",
            "user_id is null",
            "timestamp is null",
            "action_type is null",
            "search_query is null or blank for search",
            "widget_name distribution differs by action_type",
        ],
        "решение": [
            "оставить вне item graph",
            "удалить для графа",
            "удалить для sessionization",
            "удалить для последовательностей",
            "удалить для взвешенных сигналов",
            "оставить событие, но исключить из источника co-search кандидатов",
            "оставить direct item events в MVP graph, но мониторить разрез widget_name",
        ],
        "причина": [
            "search описывает намерение, а не прямое товарное взаимодействие",
            "для item-to-item пар нужны два конкретных товара",
            "нельзя построить пользовательскую историю",
            "нельзя надежно упорядочить события",
            "нельзя сопоставить событие с весом baseline",
            "текст запроса нужен для будущих query-intent сигналов",
            "контекст интерфейса может влиять на смысл клика, просмотра и добавления",
        ],
    }
)

missing_value_decisions

## 6. Анализ пользователей

Этот блок оценивает активность пользователей, чтобы найти bot-like или технический трафик. Используем полный лог, но читаем только `user_id`, `timestamp`, `action_type`, `item_id` и `widget_name`.

Отдельно считаем p99+ summary: экстремально активные пользователи должны быть помечены или ограничены в preprocessing до генерации co-visitation пар.

In [ ]:
direct_action_types = ["view", "click", "favorite", "to_cart"]
direct_item_events_lf = (
    user_actions_lf
    .filter(pl.col("action_type").is_in(direct_action_types))
    .filter(pl.col("item_id").is_not_null())
    .select(["user_id", "date", "timestamp", "action_type", "item_id", "widget_name"])
)

user_activity_lf = direct_item_events_lf.group_by("user_id").agg(
    pl.len().alias("events"),
    pl.col("item_id").n_unique().alias("unique_items"),
    pl.col("date").n_unique().alias("active_days"),
    pl.col("timestamp").min().alias("first_event_at"),
    pl.col("timestamp").max().alias("last_event_at"),
)

user_activity_summary = user_activity_lf.select(
    pl.len().alias("users"),
    pl.col("events").sum().alias("events"),
    pl.col("events").mean().alias("events_mean"),
    pl.col("events").quantile(0.5).alias("events_p50"),
    pl.col("events").quantile(0.9).alias("events_p90"),
    pl.col("events").quantile(0.95).alias("events_p95"),
    pl.col("events").quantile(0.99).alias("events_p99"),
    pl.col("events").max().alias("events_max"),
).collect()

p99_events_threshold = user_activity_summary["events_p99"].item()
total_direct_events = user_activity_summary["events"].item()

high_activity_user_summary = (
    user_activity_lf
    .filter(pl.col("events") >= p99_events_threshold)
    .select(
        pl.lit(p99_events_threshold).alias("events_p99_threshold"),
        pl.len().alias("p99_plus_users"),
        pl.col("events").sum().alias("p99_plus_events"),
        (pl.col("events").sum() / pl.lit(total_direct_events)).alias("p99_plus_event_share"),
        pl.col("events").max().alias("max_events"),
        pl.col("unique_items").max().alias("max_unique_items"),
        (pl.col("events").max() / pl.lit(total_direct_events)).alias("top_user_event_share"),
    )
    .collect()
)

top_active_users = user_activity_lf.sort("events", descending=True).limit(50).collect()

user_activity_summary, high_activity_user_summary, top_active_users

## 7. Анализ товаров, catalog reconciliation и popularity bias

Этот блок считает взвешенную популярность товаров, но не строит рекомендации. `search` по умолчанию исключается.

Поведенческое покрытие каталога считаем только через пересечение множеств `item_id`: прямое деление `behavior_items / catalog_items` не является coverage и может дать значение больше 100% при несовместимых outputs или путях данных.

In [ ]:
event_weights = load_event_weights()
item_popularity_all = weighted_item_popularity(
    user_actions_lf.select(["user_id", "timestamp", "action_type", "item_id"]),
    event_weights=event_weights,
    top_n=None,
)
item_popularity_top = item_popularity_all.head(100)

behavior_item_ids = item_popularity_all.select("item_id").unique()
catalog_item_ids = products.select("item_id").unique()

matched_items = behavior_item_ids.join(catalog_item_ids, on="item_id", how="inner").height
behavior_items_without_product = behavior_item_ids.join(
    catalog_item_ids,
    on="item_id",
    how="anti",
).height
catalog_items_without_behavior = catalog_item_ids.join(
    behavior_item_ids,
    on="item_id",
    how="anti",
).height

catalog_items = catalog_item_ids.height
behavior_items = behavior_item_ids.height

catalog_reconciliation = pl.DataFrame(
    {
        "метрика": [
            "catalog_items",
            "behavior_items",
            "matched_items",
            "behavior_items_without_product",
            "catalog_items_without_behavior",
            "catalog_behavior_coverage",
            "behavior_catalog_match_rate",
        ],
        "значение": [
            catalog_items,
            behavior_items,
            matched_items,
            behavior_items_without_product,
            catalog_items_without_behavior,
            None if catalog_items == 0 else matched_items / catalog_items,
            None if behavior_items == 0 else matched_items / behavior_items,
        ],
    }
)

catalog_reconciliation_status = pl.DataFrame(
    {
        "проверка": ["all behavior item_id values exist in product_information"],
        "статус": [
            "ok" if behavior_items_without_product == 0 else "data/path inconsistency"
        ],
        "детали": [
            "behavior_items_without_product = " + str(behavior_items_without_product)
        ],
    }
)

item_popularity_top, catalog_reconciliation, catalog_reconciliation_status

In [ ]:
popularity_decisions = pl.DataFrame(
    {
        "вопрос": [
            "Использовать raw pair_count как финальный score?",
            "Добавлять min_pair_count?",
            "Добавлять min_unique_users?",
            "Нормализовать популярность уже в baseline?",
            "Что делать с p99+ пользователями?",
            "Нужен fallback для редких и новых товаров?",
        ],
        "решение": [
            "нет",
            "да, начать с config min_pair_count=2",
            "да, обязательный quality gate первого co-visitation baseline",
            "да, cosine-like normalization в первом baseline",
            "помечать или ограничивать до генерации item pairs",
            "да, fallback по category/type/brand из товарных метаданных",
        ],
        "причина": [
            "популярные товары доминируют в сырой совместной встречаемости",
            "одиночные co-events слишком шумные",
            "один очень активный пользователь не должен определять похожесть",
            "распределение событий сильно смещено в сторону популярных товаров",
            "экстремальный пользователь может создать непропорционально много шумных пар",
            "для long-tail товаров поведенческие сигналы отсутствуют или слабы",
        ],
    }
)

popularity_decisions

## 8. Анализ поисковых запросов

`search` не входит в первый прямой item-to-item graph, но позже может стать источником co-search сигнала.

Пустые поисковые запросы считаем как `NULL` или строку, которая после `strip` становится пустой. Для топа запросов используем только непустой normalized text.

In [ ]:
search_lf = user_actions_lf.filter(pl.col("action_type") == "search").with_columns(
    pl.col("search_query").str.strip_chars().alias("search_query_trimmed")
)

search_summary = search_lf.select(
    pl.len().alias("search_events"),
    pl.col("user_id").n_unique().alias("search_users"),
    pl.when(
        pl.col("search_query_trimmed").is_not_null()
        & (pl.col("search_query_trimmed") != "")
    )
    .then(pl.col("search_query_trimmed").str.to_lowercase())
    .otherwise(None)
    .drop_nulls()
    .n_unique()
    .alias("unique_queries_normalized"),
    pl.col("search_query").null_count().alias("null_query_rows"),
    (pl.col("search_query") == "").sum().alias("empty_string_query_rows"),
    (pl.col("search_query_trimmed") == "").sum().alias("blank_query_rows"),
).collect().with_columns(
    (pl.col("null_query_rows") + pl.col("blank_query_rows")).alias("invalid_query_rows"),
    (
        (pl.col("null_query_rows") + pl.col("blank_query_rows"))
        / pl.col("search_events")
    ).alias("invalid_query_share"),
).select(
    [
        "search_events",
        "search_users",
        "unique_queries_normalized",
        "null_query_rows",
        "empty_string_query_rows",
        "blank_query_rows",
        "invalid_query_rows",
        "invalid_query_share",
    ]
)

valid_search_lf = (
    search_lf
    .filter(pl.col("search_query_trimmed").is_not_null())
    .filter(pl.col("search_query_trimmed") != "")
    .with_columns(
        pl.col("search_query_trimmed").str.to_lowercase().alias("search_query_normalized")
    )
)

top_queries = (
    valid_search_lf
    .group_by("search_query_normalized")
    .agg(pl.len().alias("rows"), pl.col("user_id").n_unique().alias("users"))
    .sort("rows", descending=True)
    .limit(50)
    .collect()
)

search_summary, top_queries

## 9. Timestamp и day-bounded возможность собрать сессии

Полная проверка session feasibility намеренно считается по дням. Так мы избегаем одного огромного глобального sort, но проверка становится day-bounded: она подтверждает корректность timestamp и форму time gaps внутри дневных партиций, а не полную глобальную sessionization через границы суток.

Production preprocessing должен сортировать события пользователя на всём выбранном периоде или явно переносить состояние сессии через date boundary.

In [14]:
session_timeout_minutes = 30
date_values = date_summary["date"].to_list()
session_summaries = []

for index, date_value in enumerate(date_values, start=1):
    print(f"[{index}/{len(date_values)}] session feasibility: {date_value}")
    day_path = ACTIONS_ROOT / f"date={date_value}"
    day_lf = (
        pl.scan_parquet(str(day_path / "**" / "*.parquet"), hive_partitioning=True)
        .filter(pl.col("action_type").is_in(direct_action_types))
        .filter(pl.col("item_id").is_not_null())
        .select(["user_id", "timestamp", "item_id"])
    )
    day_summary = time_diff_summary(
        day_lf,
        timeout_minutes=session_timeout_minutes,
        group_cols="user_id",
    ).with_columns(pl.lit(date_value).alias("date"))
    session_summaries.append(day_summary)

session_feasibility = pl.concat(session_summaries).select(
    ["date", "events", "time_diffs", "sessions", "gaps_over_timeout", "p50_seconds", "p75_seconds", "p90_seconds", "p95_seconds", "p99_seconds"]
)

session_feasibility

[1/61] session feasibility: 2024-03-01
[2/61] session feasibility: 2024-03-02
[3/61] session feasibility: 2024-03-03
[4/61] session feasibility: 2024-03-04
[5/61] session feasibility: 2024-03-05
[6/61] session feasibility: 2024-03-06
[7/61] session feasibility: 2024-03-07
[8/61] session feasibility: 2024-03-08
[9/61] session feasibility: 2024-03-09
[10/61] session feasibility: 2024-03-10
[11/61] session feasibility: 2024-03-11
[12/61] session feasibility: 2024-03-12
[13/61] session feasibility: 2024-03-13
[14/61] session feasibility: 2024-03-14
[15/61] session feasibility: 2024-03-15
[16/61] session feasibility: 2024-03-16
[17/61] session feasibility: 2024-03-17
[18/61] session feasibility: 2024-03-18
[19/61] session feasibility: 2024-03-19
[20/61] session feasibility: 2024-03-20
[21/61] session feasibility: 2024-03-21
[22/61] session feasibility: 2024-03-22
[23/61] session feasibility: 2024-03-23
[24/61] session feasibility: 2024-03-24
[25/61] session feasibility: 2024-03-25
[26/61] s

date,events,time_diffs,sessions,gaps_over_timeout,p50_seconds,p75_seconds,p90_seconds,p95_seconds,p99_seconds
str,u32,u32,i64,u32,f64,f64,f64,f64,f64
"""2024-03-01""",9853908,9696069,187884,30045,0.0,1.0,6.0,13.0,131.0
"""2024-03-02""",10811674,10642928,199706,30960,0.0,1.0,6.0,14.0,126.0
"""2024-03-03""",10502559,10331542,200334,29317,0.0,1.0,6.0,13.0,120.0
"""2024-03-04""",9100684,8940592,187757,27665,0.0,1.0,6.0,13.0,122.0
"""2024-03-05""",10270069,10097367,205132,32430,0.0,1.0,6.0,13.0,125.0
"""2024-03-06""",10910161,10749029,197408,36276,0.0,1.0,6.0,12.0,131.0
"""2024-03-07""",11567528,11394049,212155,38676,0.0,1.0,6.0,13.0,133.0
"""2024-03-08""",10351514,10188873,194679,32038,0.0,1.0,6.0,12.0,123.0
"""2024-03-09""",9440880,9292687,175138,26945,0.0,1.0,6.0,13.0,120.0


In [ ]:
session_decisions = pl.DataFrame(
    {
        "вопрос": [
            "Можно ли использовать timestamp для user timelines?",
            "Начальный session timeout",
            "Max items per session",
            "Как обрабатывать very long sessions",
            "Как обрабатывать экстремально активных пользователей",
            "Что означает текущая session_feasibility проверка",
        ],
        "решение": [
            "да, timestamp парсится и пригоден для сортировки событий",
            "30 минут",
            "начать с config max_items_per_session=50",
            "разбивать по timeout и ограничивать число товаров до генерации пар",
            "пометить p99+ пользователей и рассмотреть дневные cap-ограничения в preprocessing",
            "day-bounded diagnostic, не полная global sessionization",
        ],
        "причина": [
            "co-visitation требует упорядоченных действий пользователя",
            "стандартное MVP-значение, которое легко объяснить",
            "предотвращает комбинаторный взрыв числа пар",
            "длинные сессии смешивают разные намерения",
            "один аномальный пользователь может создать много шумных пар",
            "проверка по дням не переносит сессии через полночь",
        ],
    }
)

session_decisions

## 10. Чеклист готовности baseline и выводы EDA

Таблица становится мостиком от EDA к preprocessing и co-visitation baseline. Статусы фиксируют не только готовность, но и ограничения, которые должны быть учтены в следующем этапе.

In [ ]:
baseline_readiness = pl.DataFrame(
    {
        "проверка": [
            "данные загружены",
            "схема совпадает с контрактом",
            "timestamp корректно парсится",
            "период данных понятен",
            "дневные партиции доступны",
            "типы событий понятны",
            "типы событий для графа выбраны",
            "widget context проверен",
            "правило обработки search определено",
            "пропуски классифицированы",
            "аномалии пользователей проанализированы",
            "popularity bias проанализирован",
            "catalog-behavior reconciliation добавлен",
            "sessionization возможна",
            "можно переходить к preprocessing",
            "можно переходить к co-visitation baseline",
        ],
        "статус": [
            "готово",
            "готово",
            "готово",
            "готово",
            "готово",
            "готово",
            "готово: view/click/favorite/to_cart",
            "готово с ограничением: MVP не вводит widget-specific weights",
            "готово: search хранить отдельно для будущего co-search",
            "готово: search item_id допускается null, blank query исключается из co-search",
            "требует preprocessing decision: p99+ user cap/flag",
            "требует preprocessing decision: min_unique_users и normalization обязательны",
            "готово: coverage считается через пересечение item_id множеств",
            "готово с ограничением: текущая проверка day-bounded",
            "да, после фиксации user/session/item thresholds",
            "да, после реализации preprocessing decisions",
        ],
    }
)

baseline_readiness

## EDA conclusions

1. Product information
   - Количество товаров: 130,035.
   - Уникальных `item_id`: 130,035.
   - Дубли `item_id`: 0.
   - Пропуски в метаданных: `brand` = 227 rows, `category_name` = 89 rows; ключевые `item_id`, `name`, `type`, `category_id` без null.
   - Категория, тип и бренд можно использовать для fallback и business rules, но null brand/category_name нужно заполнять служебным значением.

2. User actions
   - Количество строк: 561,836,992.
   - Период данных: 2024-03-01 through 2024-04-30, 61 дневная партиция, 305 parquet-файлов.
   - Пользователей с direct item events: 2,948,770.
   - Direct item events: 529,562,798.
   - Catalog-behavior coverage считается только через `catalog_reconciliation`; прямое `behavior_items / catalog_items` не использовать.

3. Action types and widget context
   - Для co-visitation MVP используем direct item actions: `view`, `click`, `favorite`, `to_cart`.
   - `search` храним отдельно для будущего co-search сигнала.
   - `widget_action_profile` и `direct_widget_profile` обязательны как диагностика перед baseline; widget-specific filtering/weighting переносится на следующий этап после offline evaluation.

4. Missing values
   - `search` без `item_id` считаем допустимым и не включаем в direct item graph.
   - Товарные события без `item_id` удаляем при построении графа.
   - Критичные пропуски `user_id`, `timestamp`, `action_type` в сохранённом EDA output равны 0.
   - Search query quality: `NULL` query rows = 84,215; дополнительно учитываются blank/empty строки после `strip`, а не только null.

5. Users
   - Подозрение на bot-like или технический трафик подтверждено: max user activity = 20,309,379 direct events и 67,820 unique items.
   - p99 threshold по direct events = 2,251 events per user.
   - Top user share от direct events примерно 3.84%, поэтому p99+ users должны быть помечены или ограничены до генерации item pairs.

6. Items
   - Popularity bias есть: верхние товары имеют сотни тысяч событий и больше 100k unique users.
   - Long-tail определять через `catalog_items_without_behavior` из `catalog_reconciliation` и через будущие pair statistics.
   - Fallback нужен для rare/new/low-behavior товаров: category/type/brand с business filters.

7. Sessions
   - Рекомендуемый стартовый timeout: 30 минут.
   - Текущая `session_feasibility` является day-bounded diagnostic: она не переносит сессии через полночь.
   - Production preprocessing должен сортировать user timelines на всём периоде или явно учитывать date-boundary carry-over.
   - Ограничение длины сессии: начать с `max_items_per_session=50`.

8. Baseline decisions
   - Первый baseline: weighted co-visitation graph.
   - Score: weighted pair score + cosine-like popularity normalization.
   - Минимальные quality gates первого baseline: `min_pair_count` и обязательный `min_unique_users`.
   - До генерации пар нужны preprocessing decisions по p99+ users, session length cap и catalog reconciliation invariant.